source-snowflake/discovery.go

```diff
@@ -0,0 +1,400 @@
+func primaryKeyToCollectionKey(key string) string {
+	return "/" + escapeTildes(key)
+}
+
+func schemaFromDiscovery(info *snowflakeDiscoveryInfo) (json.RawMessage, error) {
+	// The anchor by which we'll reference the table schema.
+	var anchor = strings.Title(info.Table.Schema) + strings.Title(info.Table.Name)
+	if info.Table.Schema == "PUBLIC" {
+		anchor = strings.Title(info.Table.Name)
+	}
+
+	// Schema of the embedded "source metadata" property
+	var sourceSchema = (&jsonschema.Reflector{
+		ExpandedStruct:            true,
+		DoNotReference:            true,
+		AllowAdditionalProperties: true,
+	}).Reflect(new(snowflakeSourceMetadata))
+	sourceSchema.Version = ""
+
+	// Build `properties` schemas for each table column.
+	var properties = make(map[string]*jsonschema.Schema)
+	for _, column := range info.Columns {
+		var jsonType, err = translateDBToJSONType(column)
+		if err != nil {
+			logrus.WithFields(logrus.Fields{
+				"error": err,
+				"type":  column.DataType,
+			}).Warn("error translating column type to JSON schema")
+
+			// Logging an error from the connector is nice, but can be swallowed by `flowctl-go`.
+			// Putting an error in the generated schema is ugly, but makes the failure visible.
+			properties[column.Name] = &jsonschema.Schema{
+				Description: fmt.Sprintf("using catch-all schema: %v", err),
+			}
+		} else {
+			properties[column.Name] = jsonType
+		}
+	}
+
+	// Schema.Properties is a weird OrderedMap thing, which doesn't allow for inline
+	// literal construction. Instead, use the Schema.Extras mechanism with "properties"
+	// to generate the properties keyword with an inline map.
+	var schema = jsonschema.Schema{
+		Definitions: jsonschema.Definitions{
+			anchor: &jsonschema.Schema{
+				Type: "object",
+				Extras: map[string]interface{}{
+					"$anchor":    anchor,
+					"properties": properties,
+				},
+				Required: info.PrimaryKey,
+			},
+		},
+		AllOf: []*jsonschema.Schema{
+			{
+				Extras: map[string]interface{}{
+					"properties": map[string]*jsonschema.Schema{
+						metadataProperty: {
+							Type: "object",
+							Extras: map[string]interface{}{
+								"properties": map[string]*jsonschema.Schema{
+									"op": {
+										Enum:        []interface{}{"c", "d", "u"},
+										Description: "Change operation type: 'c' Create/Insert, 'u' Update, 'd' Delete.",
+									},
+									"source": sourceSchema,
+									"before": {
+										Ref:         "#" + anchor,
+										Description: "Record state immediately before this change was applied.",
+										Extras: map[string]interface{}{
+											"reduce": map[string]interface{}{
+												"strategy": "firstWriteWins",
+											},
+										},
+									},
+								},
+								"reduce": map[string]interface{}{
+									"strategy": "merge",
+								},
+							},
+							Required: []string{"op", "source"},
+						},
+					},
+					"reduce": map[string]interface{}{
+						"strategy": "merge",
```

Comment:
We've had issues with top-level merge reductions in some captures before, see https://github.com/estuary/connectors/pull/1144

That's mainly an issue with captures that do not have a firm schema, but I am wondering what happens with this capture if tables are altered to have columns removed, and values are not present at all for queries on the later staging tables resulting from those?

Reply:
Hmm, yeah, in that case the merged documents would end up with additional properties which are no longer in the source table. As I understand it, the reason we currently have `merge` reduction semantics on the top level of SQL CDC documents is because we need the metadata properties to be recursively merged so that `/_meta/before` will have `firstWriteWins` semantics.

It could be argued that we should revisit this at some point, but for now I think it's probably sufficient to note that this behavior for the new Snowflake capture is consistent with our other SQL CDC connectors (and in fact I think I basically just copy-pasted it from those).

Reply:
If I'm understanding correctly then, other CDC captures would do the same thing if they had a column completely removed via an alter table statement, with reduced materialized rows going forward carrying on now absent values via the merge reduction. And that's a bit of a different problem than for example dynamodb where a field may be completely absent to represent the lack of a value, where it has no concept of columns. I'd agree that it makes sense to proceed then as it is.